### DISCUSSION TITLE
Switching to .Net 6

### DISCUSSION THREAD

**Discussion by jany-tenaj** [DotSpatial/DotSpatial#1413] [Announcements]
I started transfering the code to .Net 6.

This brought up the following questions:

1. Do we still need DotSpatial.Compatibility?
It contains some old WinForms things like StatusBarPanel in IStatusBar, I can't find in .Net 6. But if I'm correct, replacing them by newer things would make us loose the backwards compatibility for which DotSpatial.Compatibility exists. 

2. Do we still need code inside DotSpatial.Positioning that seems to be for PocketPC (e.g. Device.cs)? 
If not I'd like to remove the code to clean up the project.

**Comment by sindizzy** [comment]
I use DotSpatial.Positioning functions like Normalize.

**Reply by jany-tenaj** [reply]
I don't want to get rid of DotSpatial.Positioning altogether. I just want to get rid of the [PocketPc switches](https://github.com/DotSpatial/DotSpatial/blob/master/Source/DotSpatial.Positioning/Device.cs#L119-L122) if PocketPc wouldn't be used any longer. As far as I know PocketPcs no longer exists. But I have no idea whether the code is still used or obsolete.

The way we compile DotSpatial.Positioning the PocketPc code wouldn't be used anyway. And when switching to .Net6 something was missing and the PocketPc code was active. But it wasn't compileable because of errors. That made me think that nobody was using it anyway.

**Comment by danames** [comment]
That's funny. I hadn't heard "PocketPC" for YEARS! Yeah definitely that should be able to be removed! :)

**Reply by jany-tenaj** [reply]
What about getting rid of DotSpatial.Compatibility?

**Comment by olivier57** [comment]
Hi,

Will this new version be available for .Net 5.0/x64 ?
How long will it take to upgrade it ?

Thanks for your work

Olivier

**Reply by jany-tenaj** [reply]
No I'm not putting in any effort to make this work with .Net 5.0.

**Comment by lloydchriswilson** [comment]
I have a long-term project that depends on DS that I want to upgrade to .Net 6 as well, however the reliance of DS on ContextMenu instead of ContextMenuStrip is causing problems. FYI I found that the MS UpgradeAssistant did a great job of revising my projects to work in .Net 6 pretty effortlessly. It looks like I'm losing my ability to use SOAP (which has been deprecated anyway) and ReportViewer control, but otherwise pretty good except for issues with DS. However, there may be other issues that pop up during runtime that I have not seen yet.

Anyway, **thanks** for your efforts to upgrade to .Net 6. If this can happen soon enough, I can forge ahead on my project. If it looks like it will take several months I may have to start over and just upgrade to Framework 4.8.

Let me know if there is some way I can help you with the upgrade. I'm not that familiar with the DS code, but I'm willing to try.

My two cents: you could try to put the compatibility project on the back burner and try to release test versions of .Net 6 assemblies as you get them done. My first stumbling block is DS.Controls.Legend, but I'm sure all the .Forms. projects will have similar problems.

Chris

**Reply by jany-tenaj** [reply]
@lloydchriswilson I'm doing the update because I'm thinking about upgrading my own program to .Net 6 as well.

The core dlls are already done and waiting to be pushed. The plugin dlls are giving me some trouble, because the core dlls get copied into the plugin folders although they shouldn't. >_<

**Comment by lloydchriswilson** [comment]
Thanks for your reply.

Based on what you’ve done so far, how long until a beta version will be available for download and testing? Did you create a new branch for the .Net 6 build? Are you going to give it a new version (perhaps 2.0)? Would you post to Nuget for early testing and validation, or is there some other approach you will use?

I would be doing a lot of testing of whatever you come up with. Let me know if there is anything else I can do to help.

Chris

**Comment by lloydchriswilson** [comment]
Jany,

I see you have already created a branch called MoveDsToNet6 but have not committed any updated code for a couple of months. I know you are working on the plugins but I wonder if you could upload your work to date on the various libraries understanding that they are not yet ready for prime time. I could download them and avoid reinventing the wheel as it were.

Chris

**Comment by lloydchriswilson** [comment]
@jany-tenaj , @danames , @mogikanin 

I'm glad you all uploaded 2.0.1.2 to Nuget. I've been working on trying to upgrade Version 1.9 to .Net 6 (mainly because 1.9 was the latest "Stable" version available, and it did not require me to change my calls to topology, etc.) I am making progress but it is a hard slog. Upgrade-Assistant sometimes helps a lot (after I figured out that an incorrect path in most of the .csproj files was causing problems that prevented the utility from working). I had to manually change some menu to toolstrip controls and figure out how to substitute other Nuget packages for deprecated libraries. 

I have all the core assemblies converted I think, but have not started on the plugins; therefore, the demomap project doesn't really work and I cannot test the code very thoroughly. Nevertheless, I'm making slow progress. I've actually had to do the conversion multiple times once I learned how to make the upgrade utility work better (but it still refuses to work for some projects).

My reason for doing this work is that I want to upgrade my large app that relies on DS to the best long term support library which now is .Net 6; older Frameworks will be deprecated soon. I could save a lot of effort if I just settled on Framework 4.8 for everything (which also appears to have long-term support), however published benchmark tests for .Net 6 are very promising.

I now realize that I ought to scrap everything I've done and start over with 2.0.1.2 as a base. If I do this work, I'd like to make it available to the community, however I want to know if @jany-tenaj has made any progress on his upgrade first. If not, I don't feel comfortable with the DS management issues (like version numbers, pushing updates using Git, Nuget, etc.).

If I proceed, should I connect to the MoveDStoNet6 branch or should I start a new one so I don't step on @jany-tenaj 's toes? I need someone else with more DS experience to merge my work using Git, check versioning, upload to Nuget, etc.

If I cannot get DS moved to .Net 6, then all my upgrade work on my application is wasted. I'd have to start that over and settle for Framework 4.8.

Please advise.

Chris Wilson

**Reply by jany-tenaj** [reply]
@lloydchriswilson please just wait a little bit longer until I'm finished. Fortunately I don't having anything else pressing to do currently. That means I'm able to continue updating to .Net 6. I have finished the plugins. Now I'll have to do the extensions. Then there is the tests and then I'll be done.

Basically what you wrote about DemoMap not working because of the missing plugins is the reason I haven't uploaded any changes yet. I don't like uploading code I can't test to make sure it works. ;) 

I also changed the folder structure but am not yet sure that this will be the end version. Another reason I don't want to update anything yet.

**Comment by jany-tenaj** [comment]
@lloydchriswilson If you want to have a look at the MoveDsToNet6 you'll see my changes for .Net 6. The Appveyor build isn't yet running because it can't load the nuget packages but locally this should be compileable.

**Comment by lloydchriswilson** [comment]
Please bear with me as I am new to Git and to developing a multi-developer multi-target application like this.

I just cloned the repository to my computer and selected the MoveDSToNet6 branch. I set DemoMap as my default project and ran it. There were no compiler errors but the app started without the root (map) in the legend tree, so it was not possible to add a layer to test. I then rebuilt the solution without error, but when running DemoMap I got the error "IDockManager extension must be included because a UI plugin was found." and several other similar messages and DemoMap showed an empty window.

The bin\debug\windows extensions\dotspatial.plugins.extensionmanager folder is empty. That is because that plugin is not included in the solution. The master branch has these plugins:

![image](https://user-images.githubusercontent.com/22088883/157297383-9e2ec5ec-bbe3-40f6-a30a-b560ef334b0d.png)

whereas the MoveDSToNet6 has these:

![image](https://user-images.githubusercontent.com/22088883/157297541-9ac00254-b8f0-413a-b98f-3651be1bc294.png)

Is this intentional? Should I just build the master branch and manually drop the DLLs for those missing plugins into the bin folder? Alternatively, do you want me to try to upgrade the missing plugin projects to .Net 6 myself? I'm guessing that the missing plugins were provided by the compilation of the master branch.

It seems to me that we would want each branch to output to its own unique folder, to prevent the overlap of different target outputs.

To be clear, I'm only using the assemblies found in "core" so the issues I raise may not affect my ability to create what I need and keep going, but obviously I'm concerned that I do not have my system setup correctly or I'm not using Git correctly. Running the DemoMap test app is the minimum effort I wanted to do to get started.

Next, I cleared out the bin folder, switched back to the master branch, rebuilt that (without error), and ran DemoMap. Just like the first time, DemoMap opens showing all the menus but the root item (which represents the map) is missing so layers cannot be added, and an existing .dspx file will not load without error.

On another note, may I suggest that we adopt a new version number for the .Net 6 solution. It will be easier to confirm that we are working on the correct branch, etc.

Thanks for your advice.

Chris

**Reply by lloydchriswilson** [reply]
Oops, it appears that new version is 3.0.0. Great!

**Reply by jany-tenaj** [reply]
MoveDsTonet6 will become release number 4. It is not yet released because appveyor is not yet running ;)

I have one project folder for each branch in visual studio, so I don't get in trouble with something missing or being the wrong version because of leftovers from some other branch.

In MoveDsToNet6 I have moved the extension projects from the plugin to the extensions folder, so they can be compiled to the right folder. So everything that is needed to run Demomap should be there. There were a few plugins I couldn't convert, because some classes where missing. But as I said Demomap can be run.

**Reply by lloydchriswilson** [reply]
Maybe I'm not doing this the best way, but I used VS2022 to clone the repository to my computer. When I do that, I can select which branch I want to work on. When I select the MoveDStoNet6, here is all the plugins and extensions that are added:

![image](https://user-images.githubusercontent.com/22088883/157323110-3c06aa37-0a3f-4436-87fb-3fcbd6a3039e.png)

This is strange because when I look at the github site and switch to that branch, there are other projects there that I don't see in the solution (namely, the all-important ExtensionManager extension). When I look at the .sln file online, indeed that project is missing. Is there a chance that your most recent .sln file was not uploaded? Maybe that would explain why DemoMap won't run correctly.

**Reply by jany-tenaj** [reply]
Just saw that there was a little error in MapWindowProjectFileCompatibility that only showed up after cloning to a new folder.
Pull the latest changes, clean out the bin folder, compile again. Then Demomap should run without trouble.

The projects that are missing from the solution are those that I couldn't upgrade because classes were missing or because of other reasons.

I still don't have any idea what ExtensionManager is good for. Never used it before. If you need it you're welcome to try to upgrade it.

**Reply by lloydchriswilson** [reply]
Thanks for looking at this for me. DemoMap seems to run now. I will proceed with integration of this soon-to-be version 4 and report any problems I find.

Chris

**Reply by danames** [reply]
Note: I think any code related to MapWindow Compatibility can be removed. These two projects were tied fairly closely in the early days but have gone largely different directions and I don't think this code would actually work or be needed. - Dan

**Reply by sindizzy** [reply]
> Just saw that there was a little error in MapWindowProjectFileCompatibility that only showed up after cloning to a new folder. Pull the latest changes, clean out the bin folder, compile again. Then Demomap should run without trouble.
> 
> The projects that are missing from the solution are those that I couldn't upgrade because classes were missing or because of other reasons.
> 
> I still don't have any idea what ExtensionManager is good for. Never used it before. If you need it you're welcome to try to upgrade it.

I'm fairly sure I use ExtensionManager for something. Let me look at my project and I will let you know. If so I can take a crack at converting it.

**Comment by lloydchriswilson** [comment]
All:

Here is a progress report. I am using DS for .Net 6 in my recently converted large project. I needed to make many adaptations because I had been using version 1.9 before the switch to NetTopologySuite. I also refactored my solution a lot which broke lots of things (not due to DS). 

Nevertheless, I have yet to run into any problems with the new DS release, but will be doing intensive testing in the coming weeks. I will report to everyone if I find any problems.

I noticed that there are a few updates in NuGet you might want to apply too (namely, BruTile). We need to also set a new version number, and I suggest that you generate API documentation files (which are all turned off now). Do you have a schedule for making the .Net 6 version available as beta under NuGet? 

Finally, a general .Net 6 question. Has anyone noticed and have an explanation for the following? When you create a windows forms release of an EXE, VS always creates both an EXE and DLL with the same filename. It seems like a waste.

Chris

**Reply by jany-tenaj** [reply]
We can create a nuget release as soon as I get the Appveyor errors fixed. The same goes for the new version number.

Updating BruTile might take a little more time than I have currently. But yes it has to be done.

I'll have to see when I' be able to continue because this week I'm swamped with other work.

VS2022 builds some strange dlls along the ones we need that get used for something internally.... I believe I switched those of when I still had to many errors to figure out where they were coming from. The same goes for the API documentation files. When everything runs as it's supposed to we can turn the Api documentation file generation back on.

**Comment by AlexandreDonovan** [comment]
Hi @jany-tenaj, can you tell us which branch we can use to try in our projects for .NET 6? Thanks in advance.

**Reply by jany-tenaj** [reply]
@AlexandreDonovan the master branch since yesterday evening.
I'll try to create a nuget release as soon as possible.

**Reply by AlexandreDonovan** [reply]
Thank you so much your work!

**Reply by AlexandreDonovan** [reply]
Hey I just tried installing DotSpatial.Data and I've noticed that on the nuget page, it does indeed say .NET6.0 but with Windows 7, is this normal? Also, when trying to install on a .NET 6 project, I get the follow error :
`
 Severity	Code	Description	Project	File	Line	Suppression State
Error	NU1202	Package DotSpatial.Data 4.0.656 is not compatible with net6.0 (.NETCoreApp,Version=v6.0). Package DotSpatial.Data 4.0.656 supports: net6.0-windows7.0 (.NETCoreApp,Version=v6.0)	ConsoleApp1	C:\LOCAL\Applfich\repos\VariousTests\TestNetCDFRedHat\ConsoleApp1\ConsoleApp1.csproj	1	`

I guess the windows7 thing has something to do with it. 

![Capture d’écran 2022-04-25 155901](https://user-images.githubusercontent.com/35696791/165165268-e29fa4ec-c78a-47e7-a407-943ef45440b3.png)

**Reply by jany-tenaj** [reply]
What program are you using for your tests? I created a test project in Vs2022 and could include the DotSpatial.Data from nuget without errors.

**Reply by AlexandreDonovan** [reply]
> i

I just created an empty Console Project and installed the DotSpatial.Data from Nuget package manager, version : 4.0.656 and it gives me the error mentionned in my last reply. I'm on Windows 10 and Vs2022

**Reply by stevenkangas** [reply]
I got the same error when attempting to use the nuget package in a .net6 dll library.  I could use the nuget package (without error) in a .net 6 windows forms project.

**Reply by AlexandreDonovan** [reply]
Same thing for me @stevenkangas

**Reply by jany-tenaj** [reply]
@stevenkangas @AlexandreDonovan You have to create a project with <TargetFramework>net6.0-windows</TargetFramework> because DotSpatial.Data depents on some windows things so it uses <TargetFramework>net6.0-windows</TargetFramework>. As far as I can tell those projects can't be included in net6.0 projects.

**Reply by AlexandreDonovan** [reply]
If we are not planning to use Windows Forms, should we look for Nuget package that works similarly to DotSpatial.Data? The need for us in inside a console project.

**Reply by jany-tenaj** [reply]
You might have to use something else. Basically somewhere either DotSpatial.Data or one of the referenced projects uses Forms so DotSpatial.Data has to be net6.0-windows.

**Reply by AlexandreDonovan** [reply]
Ahhh that's too bad, really wanted to use DotSpatial! Any idea what we could use instead?

**Comment by jany-tenaj** [comment]
Not really. Youcould try sharpmap but bo idea whether that works without forms.

**Comment by sindizzy** [comment]
What does this mean as usability with .NET Framework? I have many apps that still use the classic .NET DotSpatial libs. I can upgrade at some point but that will take time.

**Reply by jany-tenaj** [reply]
I'd say if you want to stay with .Net Framework you'll have to live with DS 3.0.1. I'm not planing on getting 4.x to work with .Net Framework. It's only for .Net 6.

**Comment by sindizzy** [comment]
Got it. 
Will PRs be accepted for both or just v4?

**Reply by jany-tenaj** [reply]
I wouldn't want to create another release vor v3. So accepting PRs for it might be moot?

**Reply by sindizzy** [reply]
gotcha

**Comment by lavender11** [comment]
这是来自QQ邮箱的假期自动回复邮件。
 
你好，我现在有事，无法亲自回复你的邮件。我将尽快给你回复。徐

**Reply by jany-tenaj** [reply]
Please write english. No idea who is supposed to understand you.

**Comment by denisglu** [comment]
@jany-tenaj thanks a lot for your work! 
Do you have an idea whether V4.0 release Data and Projections shall work well in .NET 7 Windows app, or are there any known obstacles?

**Comment by TongZhai** [comment]
All,

We have tools developed using the previous DS in .Net 4.5.2.
We are thinking about the long term support for .Net 4.8 and .Net 6, here is a [SO thread on this topic]( https://stackoverflow.com/questions/72413804/net-4-8-vs-net-core-net-6-regarding-lts#:~:text=NET%204.8%20is%20integrated%20in,fixes%2C%20and%20no%20new%20feature.). 

It seems .Net 6 support will end in 2024 and if MS sticks to its release rhythm, we are looking at perhaps update every couple of years, whereas .Net 4.8 will last till 2031 with Windows 11.

Without trying anything yet, my question is: can an application compiled to .Net 4.8 work (easier for now) with DS Net6 version? or do we *NEED* to update all to .NET6? (I am just trying to get some certainty before I make the plunge). Thank!!!

**Reply by TongZhai** [reply]
Just read above from @jany-tenaj about the determined path towards .NET 6+, I guess we will just go along with the ride. Thanks!

**Comment by flobadob** [comment]
We have a large project, mostly .NET Framework 4.8. I'm looking to start using DotSpatial but I'm finding that it's .net core only (since v3) and all future development will be .net core only. So, I'm reluctant to invest time in dotspatial, which is a shame. Are there any future thoughts on DotSpatial with .NET framework? Converting all our existing assemblies to .net core is not an option for us.

**Reply by flobadob** [reply]
...and I don't believe I can use a control from .net core in a .net framework control?

**Reply by sindizzy** [reply]
I have many projects at work that use the .NET Framework version and they are fine. If you find something that's broken you could always clone the repo and start your own branch with your own code.
